docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+
+  - convert ranges to page ids
+- get new page ids (not in f.pages)
+  - if have not new page ids (have been fetched)
+  - return a slice of mmaped file (f.data[start:end])
+  
+- add new page into `f.pendingPages`
+  - increase `f.pendingReaders`
+  - return `f.pendingChan`
+* Range: 
+
+  `Range` returns a slice of mmaped file (`f.data[start:end]`). 
+
+  It will not try to fetch data from obj store, and return an error if some pages in the range have not been fetched.
+
+  `Range` internals:
+
+  - convert range to page ids  
+  - if all page ids have been fetched
+    - return a slice of mmaped file
+
+  * otherwise return an error
+
+* CloseMmap: `CloseMmap` unmmaps `f.data` .
+
+* CloseFile: `CloseFile` closes  `f.file`. 
+
+* Close: `Close` closes `f.file`, unmmaps `f.data` and set `f.pages` to nil.
+
+### Range to pages
+
+Convert a range to pages:
+
+```
+const pageSize = 4 * 1024  // 4KiB
+
+minPageID := math.Floor( float(start) / float(pageSize) )
+maxPageID := math.Ceil( float(end) / float(pageSize) )
+
+pages := roaring.NewBitmap()
+pages.AddRange(minPageID, maxPageID)
+```
+
+### Combine queries
+
+```
+type BucketFile struct {
+    // ...
+    pendingDuration time.Duration   // duration to fetch pending pages
+    pendingPages    *roaring.Bitmap // pending pages to be fetched
+    pendingReaders  int             // record pending callers
+    pendingChan     chan error      // chan to notify callers the result of prefetch	
+}
+
+func (f *BucketFile) Prefetch(ranges []*Range) (chan error, error) {...}
+```
+
+BucketFile appends new pages to `pendingPages`, waits for next  time to fetch. So that we can combine multiple queries to ones.
+
+When fetching combined  pages,
+if an error occurs, BucketFile will send the error to callers via the `pendingChan` channel,
+if no error occurs, BucketFile will close the `pendingChan` channel.
+
+If callers receive an error from the channel, then prefetch failed. 
+If callers receive nil (zero value of error, after channel closed) from the channel, then prefetch done. 
+
+Each time to fetch pending pages，BucketFile will reset  `pendingPages`, `pendingReaders` and `pendingChan`.
+
+### In-memory object
+
+Each IndexReader decodes [postings offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) into memory object for sequential and  random accesses.  
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses very small memory. 
+
+For inactive IndexReaders, if the inactive duration (`now.Sub(lastQueryTime)`) greats than a configured duration (e.g. 15 mins), then IndexReader will remove in-memory objects and close file descriptor and mmap in IndexFile.  This will reduce memory usage.
+
+**NOTE**: The `pages` bitmap in BucketFile should always be kept in memory until the server is closed.
```

Comment:
Could you elaborate more on this? Why do we want this if files might disappear?

Reply:
Inactive IndexReaders just close file descriptors, not remove local files. 

Inactive IndexReaders may be queried again.
Keep `pages` bitmap in memory, so we can reuse fetched content.

Reply:
What we miss from design is human description of what `pages bitmat` is, how large it is, what the use case. Cant's see any description now. ):

Reply:
Use [Roaring bitmaps](http://roaringbitmap.org) implementation.

Compressed Size: 2.6 bits per int.

1 TiB = 1 * 1024 * 1024 * 1024 KiB = 1073741824 KiB
4 KiB pages:  1073741824 / 4 = 268435456
bitmap size: 268435456 * 3 bits = 805306368 bits = 805306368 / 8 / 1024 / 1024 = 96 MiB

so if we write 1 TiB to local files, then the size of bitmaps is 96 MiB.

Reply:
if we write many bytes (e.g. 100 TiB) to local files, bitmaps will use 10 GiB memory.

so I think we also need to close the `pages` bitmap now. ):